### PR TITLE
Fix variable lexical scoping

### PR DESCRIPTION
**Pull request by jordalgo**
This implements a `scope` stack to prevent blocks
from accessing variables outside of their lexical scope.

Example, which is no longer valid:
```
BEGIN { if (0) { $var = 1; } print($var) }
```

We need to do this for:
- While
- For
- If (both the if and else blocks)
- Unroll
- Probe
- Subprog

Issue: https://github.com/bpftrace/bpftrace/issues/3017

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Review by ajor — COMMENTED** [review]
#### AST
Can we do without having these Node types inherit from `Scope` and introducing multiple inheritance in the process?

- I'm thinking that it might make more sense for a loop to "have a" `Scope`, rather than for it to be an "is a" relationship. That'd remove the difference in implementation between `If` and the other node types
- Should this `Scope` node be called `Block` instead? Other things, e.g. a for-loop's header, can contribute to the scope but aren't part of this current `Scope` type.

#### Semantic Analyser
If `Scope` / `Block` is made into its own `Node` type, then pushing onto the `scope_stack_` member could be wrapped away in a `visit(Block&)`.

Instead of iterating over the entire scope stack each time we need to look up a variable, could we keep a cached set of valid variable names for the current scope? The cached set could be updated when entering/leaving a scope. It'd simplify looking up variables and I think also speed up semantic analysis.

#### Language Design
I'm not 100% against it, but it's the mixture of block-scoping and variables being hoisted out of if-statements that scares me. It's not clear in my mind yet what the rules are and I worry that it'll be confusing for users too. I probably need to go through some more examples.

**Comment by jordalgo** [comment]
> I'm thinking that it might make more sense for a loop to "have a" Scope, rather than for it to be an "is a" relationship.

Yeah, I like this. Good call. Will update.

> Should this Scope node be called Block instead? 

I went back and forth on this and TBH I'm not sure what the correct answer is. Scope seems more explicit to me and self-explanatory, especially if we include it as a member of for, while, if, etc... I like saying "While statements have a Scope" rather than "While statements have a block".

> If Scope / Block is made into its own Node type, then pushing onto the scope_stack_ member could be wrapped away in a visit(Block&).

But what else would this node do? We still have to explicitly `pop` outside this node's `visit` right? 

> Instead of iterating over the entire scope stack each time we need to look up a variable, could we keep a cached set of valid variable names for the current scope? The cached set could be updated when entering/leaving a scope. It'd simplify looking up variables and I think also speed up semantic analysis.

Maybe we can amortize this in some way but yeah I agree with the sentiment.

BTW, I'll probably hold off on changes until we resolve https://github.com/bpftrace/bpftrace/issues/3017

**Comment by ajor** [comment]
> > Should this Scope node be called Block instead?
> 
> I went back and forth on this and TBH I'm not sure what the correct answer is. Scope seems more explicit to me and self-explanatory, especially if we include it as a member of for, while, if, etc... I like saying "While statements have a Scope" rather than "While statements have a block".

I think `scope == block` for most block-types, but not for-loops or functions.
```
for ($kv : @map) { ... }
fn add($a: int64): int64 { ... }
```
`$kv` and `$a` are both in the scope of the above nodes, but not in their blocks.

> > If Scope / Block is made into its own Node type, then pushing onto the scope_stack_ member could be wrapped away in a visit(Block&).
> 
> But what else would this node do? We still have to explicitly `pop` outside this node's `visit` right?

I was imagining something like this:
```
visit(Scope &scope)
{
  scope_stack_.push_back(scope);
  for (Statement &stmt : scope.stmts) {
    accept(stmt);
  }
  scope_stack_.pop_back();
}
```

It'd also come of more use to centralise the logic for caching of valid variables in the current scope if we added that optimisation.

**Comment by jordalgo** [comment]
```
visit(Scope &scope)
{
  scope_stack_.push_back(scope);
  for (Statement &stmt : scope.stmts) {
    accept(stmt);
  }
  scope_stack_.pop_back();
}
```
Looking again at the individual places where we push on to the scope_stack, there is logic specific to the block nodes that happens in between the push and pop e.g.
```
void SemanticAnalyser::visit(While &while_block)
{
  scope_stack_.push_back(&while_block.scope);

  if (is_final_pass() && !bpftrace_.feature_->has_loop()) {
    LOG(WARNING, while_block.loc, out_)
        << "Kernel does not support bounded loops. Depending"
           " on LLVMs loop unroll to generate loadable code.";
  }

  while_block.cond->accept(*this);

  loop_depth_++;
  accept_statements(while_block.scope.stmts);
  loop_depth_--;
  scope_stack_.pop_back();
}
```
So I'm not sure how we can easily make `Scope` a node without bloating it.

**Comment by ajor** [comment]
That could just become this, no?
```
void SemanticAnalyser::visit(While &while_block)
{
  if (is_final_pass() && !bpftrace_.feature_->has_loop()) {
    LOG(WARNING, while_block.loc, out_)
        << "Kernel does not support bounded loops. Depending"
           " on LLVMs loop unroll to generate loadable code.";
  }

  while_block.cond->accept(*this);

  loop_depth_++;
  while_block.scope->accept(*this);
  loop_depth_--;
}
```
I don't think there's any need for the warning and loop condition to occur within the scope of the loop's body.

**Comment by jordalgo** [comment]
> I don't think there's any need for the warning and loop condition to occur within the scope of the loop's body.

I think the condition *should* exist within the scope of the loop's body. Let's look at a `for` loop for example:
```
for ($kv : @map) {
  print($kv.0); // key
  print($kv.1); // value
}
```
`$kv` should not be accessible outside of the for loop. I believe this is true for other languages as well though C++ directly comes to mind e.g. these are both errors
```
std::vector<int> my_vec{1, 2};
for (auto i : my_vec) {
  std::cout << i;
}
std::cout << i; // undefined error
```
```
if (int i = some_condition()) {
  ...
}
std::cout << i; // undefined error
```

**Comment by jordalgo** [comment]
> Instead of iterating over the entire scope stack each time we need to look up a variable, could we keep a cached set of valid variable names for the current scope? The cached set could be updated when entering/leaving a scope. It'd simplify looking up variables and I think also speed up semantic analysis.

@ajor I looked into this but I don't think there is any additional optimization here. At the time of scope iteration, we're only already iterating over eligible scopes and all the variables in those are valid. Also the scopes are all unique and are looking up maps of variables. Let me know if I'm missing something obvious.

**Comment by ajor** [comment]
> I think the condition _should_ exist within the scope of the loop's body. Let's look at a `for` loop for example:

Agreed about for-loops, and that's why I think `Scope` is a misleading name for this node! The node doesn't include a for-loop's header, despite the header forming part of the lexical scope.

I still think it should be a proper AST `Node` type though. The way the AST is built is that everything is either a `Node` or a list of nodes and it would be odd to violate that for one particular type. If we're struggling to make it fit, then maybe we've got the wrong abstractions.

What about having a separate, non-AST data type to represent the abstract lexical scope. Then the AST node can just be a `Block`, which most of the time is directly used to build a `Scope,` but not always (e.g. in for-loops).

**Comment by ajor** [comment]
> @ajor I looked into this but I don't think there is any additional optimization here. At the time of scope iteration, we're only already iterating over eligible scopes and all the variables in those are valid. Also the scopes are all unique and are looking up maps of variables. Let me know if I'm missing something obvious.

In this script:
```
$a = 1;
{
  $b = 1;
  {
    $c = 1;
    {
      $d = $1;
      print($d);
    }
  }
}
```

The scope stack will look something like this:
```
{$a}
{$b}
{$c}
{$d}
```

When we come to the `print($d)` statement, the current implementation will iterate over each level of the scope stack and perform a map-lookup in each one until it finds `$d`. That makes it an `O(s)` operation where `s` is the size of the scope stack.

What I'm suggesting is to instead build the scope stack like this, where all live variables are present in the top level of the stack:
```
{$a}
{$a,$b}
{$a,$b,$c}
{$a,$b,$c,$d}
```

Then the `print($d)` statement only ever needs to do one map-lookup into the top level of the scope stack - an `O(1)` operation.

**Comment by jordalgo** [comment]
> When we come to the print($d) statement, the current implementation will iterate over each level of the scope stack and perform a map-lookup in each one until it finds $d. That makes it an O(s) operation where s is the size of the scope stack.

Right but I don't think we have any real life examples where the scope stack is ever bigger than 10 (and that's being generous).

> What I'm suggesting is to instead build the scope stack like this, where all live variables are present in the top level of the stack:

In this case we're paying for operations up front (that is the inserting, or removing, of references, or copies, of the variables into a new set/map when we create a new scope) when we're not sure we're going to need them.

**Comment by jordalgo** [comment]
> Agreed about for-loops, and that's why I think Scope is a misleading name for this node! 

Still not sure if I understand why the name is misleading for this example but I understand that if you want to see this in the AST then it probably reads better as Block.

> What about having a separate, non-AST data type to represent the abstract lexical scope. Then the AST node can just be a Block, which most of the time is directly used to build a Scope, but not always (e.g. in for-loops).

I'm not sure if I understand what you mean here. Do you mind adding an example? I think I'm having trouble here because I don't see a easy way to generalize the logic that exists between pushing a scope to the stack and popping it off. Or do you mean, we can have a `BlockStart` and `BlockEnd` node that the parent Node can `visit` at different logic points that controls the scope?

**Comment by jordalgo** [comment]
@ajor and I cleared some things up offline. Going to put up a few minor changes and then I think this should be good to go.

**Comment by jordalgo** [comment]
Please vote here if you want this diff to move forward. *Note*: we will first tally the votes from only maintainers and very active contributors but also take into account votes from the larger community to make this decision.

- **Thumbs up** for this PR (or some version of it): **block/lexical scoping** 👍 
- **Thumbs down** for keeping things as they are: **Probe scoping** 👎 
- Any other random emoji if you don't feel strongly.

File: tools/tcpdrop.bt
Comment (viktormalik) [inline tools/tcpdrop.bt:85]: While I'm a fan of block scoping in general, the fact that `$daddr` is not accessible after the `if`/`else` block is quite a big backwards compatibility breakage. I can imagine a lot of scripts out there which rely on this functionality.

I may have missed something in the discussions but do we plan to introduce variable declarations so that the only update people need to do is add an explicit declaration before the condition? Or do we even consider declaration hoisting?

Comment (fbs) [reply tools/tcpdrop.bt:85]: This was discussed on IRC a bit but I dont think anything came out of it.

I'm not a fan of hoisting as it basically undoes the scoping. You're lifting the var to the outer scope, might as well stick with probe scope. Nested if statements also become confusing

If we do block scoping we also need a way to do upfront declaration. Either the ugly way :

```
$daddr = ntop(dummy)
if ...
  $daddr = ntop(real)
 ```
 
 or by adding the option to define upfront.
 
 This probably shouldn't land until a decision has been made here so good to bring it up

Comment (jordalgo) [reply tools/tcpdrop.bt:85]: I think the general consensus is that we want to introduce variable declarations, which seem to coincide with both the [user-defined functions PR](https://github.com/bpftrace/bpftrace/pull/3068) and this [map declarations RFC](https://github.com/bpftrace/bpftrace/issues/2954). If we land this PR without waiting for variable declarations then to @fbs 's point, users have to use the ugly way in that they need to initialize with a junk or 0 value. 

IMHO, variable declarations don't seem very controversial and are probably not that hard to implement, so I'm fine doing that work either stacked on top of this PR or to follow closely after before we release another bpftrace version with this in it.

Comment (viktormalik) [reply tools/tcpdrop.bt:85]: So, to make things clear, with variable declarations but no hoisting (which is roughly the plan), the above will become
```
let $daddr, $saddr: inet_t;
if ($inet_family == AF_INET) {
  $daddr = ntop($sk->__sk_common.skc_daddr);
  $saddr = ntop($sk->__sk_common.skc_rcv_saddr);
} else {
  $daddr = ntop($sk->__sk_common.skc_v6_daddr.in6_u.u6_addr8);
  $saddr = ntop($sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr8);
}
printf("%39s:%-6d %39s:%-6d %-10s\n", $saddr, $lport, $daddr, $dport, $statestr);
```
and semantic analyser will check that the value is defined in all branches, otherwise print a warning.

From users perspective, they will just need to add explicit declarations to their scripts. It's still a backwards compat breakage but certainly something we can live with.

Comment (jordalgo) [reply tools/tcpdrop.bt:85]: @viktormalik That's pretty much how I imagine it - maybe without the `let` keyword 😉

Comment (jordalgo) [reply tools/tcpdrop.bt:85]: @fbs @viktormalik I've got a first draft of variable declarations (still needs more tests and a fix that hasn't been merged yet) just to give you an idea of what it looks like. It's pretty simple, which is good:
https://github.com/jordalgo/bpftrace/commit/2bcb96a35287e39fb20e12cfb121a1e334474aab

Comment (viktormalik) [reply tools/tcpdrop.bt:85]: Looks pretty straightforward, nice!

Comment (fbs) [reply tools/tcpdrop.bt:85]: IIRC some other MR/proposal added a let syntax. Should we centralize that work in an rfc like issue and block this MR on that?

Comment (jordalgo) [reply tools/tcpdrop.bt:85]: @fbs Are you talking about the map declarations proposal? https://github.com/bpftrace/bpftrace/issues/2954 (cc @JakeHillion)

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2979]: How come this condition was not necessary before?

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2994]: Why do this specifically on the first and the last pass? Can we add a comment to explain?

File: src/ast/passes/semantic_analyser.h
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.h:162]: Representing the scope stack with `std::vector` means that we're always iterating it in the "opposite" order, i.e. the outermost scope first. It shouldn't matter now as we don't support variable shadowing but if we ever decide to do so, this may become a problem so I'd at least explicitly state the fact here.

File: tests/semantic_analyser.cpp
Comment (viktormalik) [inline tests/semantic_analyser.cpp:4322]: I know it's just a test but could we have a nicer indentation of programs which span multiple lines? Same for the other test cases in this block.

**Review by viktormalik — COMMENTED** [review]
I like how straightforward this change mostly is. Logic looks good to me, I just have a few code readability remarks.

File: man/adoc/bpftrace.adoc
Comment (jordalgo) [reply man/adoc/bpftrace.adoc:941]: Cause they won't always be -> all of @ryantimwilson's work to sometimes put things in a global scratch buffer. It seems like an implementation detail that users don't need to know about.

File: src/ast/passes/semantic_analyser.h
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.h:162]: Good call.

File: man/adoc/bpftrace.adoc
Comment (viktormalik) [reply man/adoc/bpftrace.adoc:941]: Yeah, but I thought that at the moment, scratch variables are always stored on stack. My point is that I wouldn't change this in the docs before the actual implementation changes.



> It seems like an implementation detail that users don't need to know about.

I don't think it's an implementation detail. It inherently limits the overall size of scratch variables which is something that users can run into.

Comment (jordalgo) [reply man/adoc/bpftrace.adoc:941]: kk, I'll change it back for now.

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:2979]: Since variables are no longer probe scoped. This is now valid `BEGIN { if (1) { let $x; } else { let $x; } let $x; }` - See how we define the last $x in the outer scope at the end. Now on subsequent passes, it's viewed as variable shadowing in the if/else blocks.

Comment (viktormalik) [reply src/ast/passes/semantic_analyser.cpp:2979]: Right, makes sense. A comment would be helpful, though.

File: src/ast/passes/semantic_analyser.cpp
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:2994]: The only error checks below are for last or first pass so no need to iterate over the scope stack in other passes. I'm adding a comment as to why we check for `first_pass` - it's similar to above. For cases like this: `BEGIN { if (1) { let $x; } $x = 2; }`

**Comment by jordalgo** [comment]
@viktormalik I think I addressed all your comments. 🤞

File: man/adoc/bpftrace.adoc
Comment (viktormalik) [inline man/adoc/bpftrace.adoc:964]: Should we say that it will be 0 if the condition was not true?

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2392]: Do we still need to pass `&f`?

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2988]: typo?
```suggestion
      // Again, this is legal and there is no ambiguity but `$x = 2` gets
```

**Review by viktormalik — COMMENTED** [review]
> @viktormalik I think I addressed all your comments. 🤞

You did, I just found 3 very small things :-)

**Comment by jordalgo** [comment]
@viktormalik with the sharp eyes 👀

**Comment by jordalgo** [comment]
Fixed 👍

**Review by viktormalik — APPROVED** [review]
Nice!

**Comment by jordalgo** [comment]
@fbs Any objections to merging?

**Pull request merged**
